pubsub/google/cloud/pubsub_v1/publisher/batch/thread.py

```diff
@@ -0,0 +1,236 @@
+
+        This moves the batch out from being the active batch to an in-flight
+        batch on the publisher, and then the batch is discarded upon
+        completion.
+
+        .. note::
+
+            This method blocks. The :meth:`commit` method is the non-blocking
+            version, which calls this one.
+        """
+        # Update the status.
+        self._status = 'in-flight'
+
+        # Sanity check: If there are no messages, no-op.
+        if len(self._messages) == 0:
+            return
+
+        # Begin the request to publish these messages.
+        response = self.client.api.publish(
+            self._topic,
+            self.messages,
+        )
+
+        # We got a response from Pub/Sub; denote that we are processing.
+        self._status = 'processing results'
+
+        # Sanity check: If the number of message IDs is not equal to the
+        # number of futures I have, then something went wrong.
+        if len(response.message_ids) != len(self._futures):
+            for future in self._futures:
+                future.set_exception(exceptions.PublishError(
+                    'Some messages were not successfully published.',
+                ))
+            return
+
+        # Iterate over the futures on the queue and return the response IDs.
+        # We are trusting that there is a 1:1 mapping, and raise an exception
+        # if not.
+        self._status = self.Status.SUCCESS
+        for message_id, future in zip(response.message_ids, self._futures):
+            future.set_result(message_id)
+
+    def monitor(self):
+        """Commit this batch after sufficient time has elapsed.
+
+        This simply sleeps for ``self._settings.max_latency`` seconds,
+        and then calls commit unless the batch has already been committed.
+        """
+        # Note: This thread blocks; it is up to the calling code to call it
+        # in a separate thread.
+        #
+        # Sleep for however long we should be waiting.
+        time.sleep(self._settings.max_latency)
+
+        # If, in the intervening period, the batch started to be committed,
+        # then no-op at this point.
+        if self._status != self.Status.ACCEPTING_MESSAGES:
+            return
+
+        # Commit.
+        return self._commit()
+
+    def publish(self, message):
+        """Publish a single message.
+
+        Add the given message to this object; this will cause it to be
+        published once the batch either has enough messages or a sufficient
+        period of time has elapsed.
+
+        This method is called by :meth:`~.PublisherClient.publish`.
+
+        Args:
+            message (~.pubsub_v1.types.PubsubMessage): The Pub/Sub message.
+
+        Returns:
+            ~.pubsub_v1.publisher.futures.Future: An object conforming to
+                the :class:`concurrent.futures.Future` interface.
+        """
+        # Coerce the type, just in case.
```

Comment:
> Basically everywhere that we expect protobuf objects, we accept dictionaries.

That isn't remotely true, we have to draw the line somewhere.

Bleh, I don't like this but I don't feel like I can talk you out of it. It can stay as-is.